### PR TITLE
Fix block detail updating

### PR DESCRIPTION
**Pull request by AtkinsChang**
In ethcore client, update of `block detail` occurs in two locations: `BlockChain::mark_finalized` and `BlockChain::insert_block`
https://github.com/paritytech/parity-ethereum/blob/a9cb5722383cb89ebf38ba4ea2fa2073f15d5f8b/ethcore/src/client/client.rs#L527
https://github.com/paritytech/parity-ethereum/blob/a9cb5722383cb89ebf38ba4ea2fa2073f15d5f8b/ethcore/src/client/client.rs#L536-L539
But in `BlockChain::prepare_block_details_update` (called by insert_block), parent block detail is read from cache/db, this may cause previous modification of `BlockChain::mark_finalized` be overwritten.
https://github.com/paritytech/parity-ethereum/blob/a9cb5722383cb89ebf38ba4ea2fa2073f15d5f8b/ethcore/blockchain/src/blockchain.rs#L1345-L1348

testcase is added to reproduce this situation

### CODE REVIEW

**Comment by AtkinsChang** [comment]
`mark_finalized` updates the `pending_block_details`
`prepare_block_details_update ` reads from `block_details` so the previous `is_finalized` update is overwritten
try the testcase in second commit 784e159

Handling parent is the special case that it reads data updated previously in the same transaction so I use this.

I can change to another solution like add a 2-layered cache read and replace
https://github.com/paritytech/parity-ethereum/blob/a9cb5722383cb89ebf38ba4ea2fa2073f15d5f8b/ethcore/blockchain/src/blockchain.rs#L1347
with 
```rust
let mut parent_details = self.db.key_value().read_with_two_layer_cache(db::COL_EXTRA, &self.pending_block_details, &self.block_details, hash)
```
if you prefer this

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -536,6 +540,7 @@ impl Importer {
 		let route = chain.insert_block(&mut batch, block_data, receipts.clone(), ExtrasInsert {
 			fork_choice,
 			is_finalized,
+                        is_parent_finalized,
```
[the remark this answers is not in the document]
Comment (AtkinsChang) [reply ethcore/src/client/client.rs]: fixed in force-pushed commit

**Comment by tomusdrw** [comment]
> mark_finalized updates the pending_block_details
prepare_block_details_update reads from block_details so the previous is_finalized update is overwritten

I see now, thanks for explanations.

The `Blockchain` API indeed seems very fragile, I feel we can run into more issues if we commit multiple block at once in a single batch (see `insert_block`) - all the updates to `block_details` will only end up in `pending_block_details` cache, not `block_details` cache, so we will just keep overwriting the same stuff.

Imagine we are importing a `A -> B` and `A -> B'` in single `batch`, we should end up with two `children` entries in `A`, but we will only have one.

I think I'm more inclined to the double cache approach, and go through all methods that work on a `batch` to make sure we are always taking `pending_blocks_details` cache into account whenever we are preparing the `ExtrasUpdate`.
In prinicple there is another issue, where you have two concurrent `batch`es present - the blockchain API doesn't really prevent that, but I think we never do that in practice (so I don't think we need to address it here).

Do you also feel the issues I've raised above are valid?

**Comment by AtkinsChang** [comment]
Yup, it may cause data inconsistency for two concurrent batch.

And aside from this issue, there are also issues for warp sync that cause wrong block details.

**Comment by AtkinsChang** [comment]
@tomusdrw Updated as your suggestion

**Comment by dvdplm** [comment]
@AtkinsChang good stuff, thanks! Before merging we're going to run a few tests from this branch just to make sure we don't mess up. 
/cc @adilimroz

**Comment by adilimroz** [comment]
@dvdplm Ran syncs with the mentioned commits. It all looks good to me.

File: ethcore/blockchain/src/blockchain.rs [hunk 2]
```diff
@@ -195,6 +195,12 @@ pub trait BlockProvider {
 		where F: Fn(&LogEntry) -> bool + Send + Sync, Self: Sized;
 }
 
+/// Interface for querying blocks by hash and by number with pending transaction.
+trait InTransactionBlockProvider {
+	/// Get the familial details concerning a block.
```
[the remark this answers is not in the document]
Comment (AtkinsChang) [reply ethcore/blockchain/src/blockchain.rs]: This doc comes from
https://github.com/paritytech/parity-ethereum/blob/a9cb5722383cb89ebf38ba4ea2fa2073f15d5f8b/ethcore/blockchain/src/blockchain.rs#L124

Imo the block details store finalize state and children hash, so I think "familial" word is proper here

**Comment by AtkinsChang** [comment]
Thanks for review
fixed in amended commit

**Pull request merged**